### PR TITLE
[Code Health] clang-tidy cleanup, part 1 

### PR DESCRIPTION
**Pull request by msiddhu**
Contributes to #2053 

## Changes

Please provide a brief description of the changes here.

Added the `.clang-tidy` configuration file. Solved some warnings using auto FIX-IT and some manually. [Suppressed some warnings](https://clang.llvm.org/extra/clang-tidy/#suppressing-undesired-diagnostics) where needed (mostly in test cases, headers, and recursive functions).

Started with configurations that are a good starting point, inspired by [Google Cloud clang-tidy](https://github.com/googleapis/google-cloud-cpp/blob/main/.clang-tidy) and [AWS Palace clang-tidy](https://github.com/awslabs/palace/blob/main/.clang-tidy).

More checks could be enabled in further cleanups.

### Points to note:
1. A lot of _pass by value_ to _const pass by reference_.
2.  Addition and removal of _rvalue_ references where ever possible and needed.
3. String format changes such as  `.append` instead of `+`  ,  `'\n'` instead of `std::endl` , added missed comma's
4. Removing `using namespace` which are not used in source code.
5. Added exceptions  where it was needed and clang-tidy warnings where thrown.
6. Changing from recursive to iterative ( metrics circular buffer, for the other function (json) used suppress warnings)
7. `long` to `int64` [Google Coding conventions](https://google.github.io/styleguide/cppguide.html#Integer_Types)

### Cleanup in following Areas:

1. API: tests 
2. EXAMPLES  
3. EXPORTERS
     -   Elasticsearch
     -   Memory
     -   OStream
     -   OTLP
     -   Prometheus
     -   Zipkin
4. HTTP Client and its tests
5. SDK: Traces, Logs, Metrics and their tests.

Please review the differences in indented block as they'll be shown as complete block making it difficult to find differences. Hopefully this cleanup may provide some performance improvements.

## Remaining Work:

More cleanup, Integrate in CI .

For significant contributions please make sure you have completed the following items:

* [ ] `CHANGELOG.md` updated for non-trivial changes
* [ ] Unit tests have been added
* [ ] Changes in public API reviewed

### CODE REVIEW

**Comment by lalitb** [comment]
@msiddhu Can you update the description with the rationale behind these changes? Seems some of the changes in the API could result in ABI incompatibility.

**Comment by msiddhu** [comment]
I'm trying to add clang-tidy into CI. Creating a cleanup commit. I'm having tough time deciding what clang-tidy checks should be present. We'll be back in a couple of days after testing it.

**Comment by marcalff** [comment]
Thanks for the cleanup.

Reopening and setting as draft, we definitively don't want this work to be forgotten ;-)

**Comment by marcalff** [comment]
@msiddhu Thanks for working on this, this cleanup is much needed.

From experience in similar cleanup for warnings:

* https://github.com/open-telemetry/opentelemetry-cpp/pulls?q=is%3Apr+is%3Aclosed+author%3Amarcalff+warning

the complexity in resolving the cleanup is not so much individual changes, but making sure the overall task makes progress, including with reviews and merges.

Several activities are needed:

* fix existing warnings, sure,
* deliver fixes in workable patches, that can be reviewed,
* implement some CI to detect warnings in the code.

The clang-tidy cleanup will be an iterative process, the best is to do it by small size chunks, instead in one massive PR.

Please proceed as follows:

* (1), **do not ever use force-push**. This destroys the previous history on the PR, and resets the review to the very beginning. Do several commits instead, which is easier to review, and also easier to rollback if needed. The number of commits does not matter (see #2518), because the PR will be merged with a squash at the end.

* (2), this PR has some merge conflicts, because another cleanup is already in progress, with include-what-you-use, which also touches a lot of files. Please resolve the existing merge conflicts for this PR (it should be trivial). We can hold on on iwyu to avoid generating more conflicts for a while, to give time to clang-tidy to progress.

* (3), and it is a very weird thing to say: the content of the `.clang-tidy` file (which warning to report) is not important ... at this point. The starting list of settings looks good. What is more important is to merge the `.clang-tidy` file and have some CI to report warnings, to get the process to bootstrap. Once warnings are reported in CI, subsequent PRs can fix them, and the list of warnings to enforce can grow other time.

* (4), as @lalitb already pointed out, the content of `api/*` is very sensitive, and will take much longer to review. Because of this, changes touching `api/*` are better done in a separate PR, to avoid blocking progress for everything else. Note that changes to the api tests, as in `api/test/*` are ok, that part is not sensitive to ABI breaks, so just keep the fixes already implemented there.

* (5), reviewers need some time to look at changes, so the PR should not constantly change with more fixes, as it delays the review ever more. Since this PR already touches 100 files and 1K LOC, this is a good chunk to look at already, so please do not add more fixes to "part 1". Once part 1 is merged, more fixes can be implemented in a different PR of course.

From a quick look at the PR current content, the review should be fast for this, so I think we can aim to merge this chunk soon. Please resolve the merge conflicts and flag the PR as ready to review.

Again, thanks for taking this up.

**Comment by marcalff** [comment]
Not sure why CI is stuck. First thing is to resolve conflicts anyway, the next commit should restart CI.

**Comment by msiddhu** [comment]
@marcalff 
I think the hard CI check using the no-exceptions flag is failing. Please help me with this. What could the solutions be?

**Comment by lalitb** [comment]
>I think the hard CI check using the no-exceptions flag is failing. Please help me with this. What could the solutions be?

Was there any need to add try-catch for [OTELResourceDetector::Detect()](https://github.com/msiddhu/opentelemetry-cpp/blob/2825d0372bd26146f0172801851c53b57fbd726f/sdk/src/resource/resource_detector.cc#L66)? The current CI is strict enough not to allow adding exception in the code.

**Comment by lalitb** [comment]
@msiddhu - As you are more familiar with this,  checking your opinion on  how easy it would to modify the code to use the safe, non-throwing alternatives. I believe most of the error flagged by clang-tidy for missing exception handling is coming from C++ Standard Library ? Not asking you to do any change here, just trying to understand the complexity. :)

**Comment by msiddhu** [comment]
Hello @lalitb , 

I feel that would be complex avoiding exceptions and using non-throwing alternatives.

For example, one of the functions got flagged:
```
/home/msiddhu/opentelemetry-cpp/exporters/elasticsearch/src/es_log_recordable.cc:238:31: warning: an exception may be thrown in function 'SetTraceId' which should not throw exceptions [bugprone-exception-escape]
  238 | void ElasticSearchRecordable::SetTraceId(const opentelemetry::trace::TraceId &trace_id) noexcept
```
[ElasticSearchRecordable::SetTraceId](https://github.com/open-telemetry/opentelemetry-cpp/blob/6dbfdb52ebca82d8705392fe88dadd9c133374e4/exporters/elasticsearch/src/es_log_recordable.cc#L238) has been flagged. 

In this particular function, the **string constructor** may produce a **bad allocation exception** due to no memory being available (which is very rare):
`std::string(trace_buf, sizeof(trace_buf)); `

In some cases, we could use other alternatives, but in others, we couldn't, in my opinion.

The attached file contains the **clang-tidy flagging related to exceptions**, which I have not yet addressed as the size of the PR is already large.

[function_exceptions.txt](https://github.com/user-attachments/files/16133946/function_exceptions.txt.txt)

Just check out what functions are flagged by clang-tidy to get an idea.

**Comment by marcalff** [comment]
>     3. Noted. I had a doubt about the wanted to ask about CI, The CI will only run on the files which are touched in the particular PR right? Because running clang-tidy on every file(full project except some components) in every commit will take very long time I feel. At Least for me its taking long time to run clang-tidy.
> 

CI will run on every file, in every PR, on every commit.

How long is a full clang-tidy run taking currently, and how does it compare to a regular build ?

Runners can (and should) run clang-tidy in parallel (`make -j $(nproc)`) when possible, to reduce the total latency.

>     4. Noted. For this PR I removed the changes in API. Will create a new PR with those changes soon.
> 
>     5. Noted.
> 
> 
> #### Question: Should the CI for clang-tidy be provided in this PR or new PR for that?
> 

Please do this in a different PR. See notes in #2051

> Still one check is failing (Bazel noexcept). I'll resolve that soon and make the PR ready to review.
> 
> Please let me know if anything is needed. Thanks.

Starting the review now.

File: examples/etw_threads/main.cc
Comment (marcalff) [inline examples/etw_threads/main.cc:116]: Why ?

If the issue is that argc and argv are not used, just comment the parameter names, but do not change the prototype of main.

File: exporters/ostream/test/ostream_log_test.cc
Comment (marcalff) [inline exporters/ostream/test/ostream_log_test.cc:108]: Works too, but the original code was correct: string concatenation done in the parser.

File: exporters/otlp/test/otlp_http_exporter_test.cc
Comment (marcalff) [inline exporters/otlp/test/otlp_http_exporter_test.cc:161]: Why is the code formatted differently ?

All the formatting is already checked in CI, with clang-format.

Not sure if/why clang-tidy changed it, or if this is the result of running a possible different version of clang-format, but it will be better if clang-tidy avoids touching the code format at all.

We can't have multiple tools compete on this, generating noisy diffs.

File: exporters/prometheus/test/exporter_utils_test.cc
Comment (marcalff) [inline exporters/prometheus/test/exporter_utils_test.cc:107]: The case conveys semantic, it shows the UnTyped case was not forgotten.

Ok to remove the break and fall thru the default, but not ok to remove the case entirely.

File: exporters/zipkin/src/zipkin_exporter.cc
Comment (marcalff) [inline exporters/zipkin/src/zipkin_exporter.cc:78]: This is a try/catch in the production code, which changes the logic.

The try is on each span in the loop, it could have been on the entire export too.

Not saying this is correct or incorrect, but a massive cleanup that contains 1000 lines of changes can not contain logic changes like this buried in the middle of it.

How to deal with exceptions to comply with noexcept will need more analysis, and should be better done separately.

File: ext/include/opentelemetry/ext/http/client/curl/http_operation_curl.h
Comment (marcalff) [inline ext/include/opentelemetry/ext/http/client/curl/http_operation_curl.h:271]: I think this will break curl on 32bits platforms.

Curl does use `long` in the API, not int64_t.

Please revert.

File: ext/include/opentelemetry/ext/http/client/curl/http_operation_curl.h
Comment (marcalff) [inline ext/include/opentelemetry/ext/http/client/curl/http_operation_curl.h:304]: This relates to CURLINFO_RESPONSE_CODE, which uses a long.

Revert.

File: ext/src/http/client/curl/http_operation_curl.cc
Comment (marcalff) [inline ext/src/http/client/curl/http_operation_curl.cc:456]: Revert the long -> int64_t.
Keep the const reference.

File: ext/src/http/client/curl/http_operation_curl.cc
Comment (marcalff) [inline ext/src/http/client/curl/http_operation_curl.cc:473]: Revert the long -> int64_t.
Keep the const reference.

File: ext/src/http/client/curl/http_operation_curl.cc
Comment (marcalff) [inline ext/src/http/client/curl/http_operation_curl.cc:532]: Revert

File: ext/src/http/client/curl/http_operation_curl.cc
Comment (marcalff) [inline ext/src/http/client/curl/http_operation_curl.cc:732]: Revert

File: ext/src/http/client/curl/http_operation_curl.cc
Comment (marcalff) [inline ext/src/http/client/curl/http_operation_curl.cc:734]: Revert

File: ext/src/http/client/curl/http_operation_curl.cc
Comment (marcalff) [inline ext/src/http/client/curl/http_operation_curl.cc:758]: Revert

File: ext/src/http/client/curl/http_operation_curl.cc
Comment (marcalff) [inline ext/src/http/client/curl/http_operation_curl.cc:776]: Revert

File: ext/src/http/client/curl/http_operation_curl.cc
Comment (marcalff) [inline ext/src/http/client/curl/http_operation_curl.cc:822]: Revert

File: ext/src/http/client/curl/http_operation_curl.cc
Comment (marcalff) [inline ext/src/http/client/curl/http_operation_curl.cc:843]: Revert

File: sdk/src/logs/logger.cc
Comment (marcalff) [inline sdk/src/logs/logger.cc:62]: Returning a nullptr is likely to break things.

Please revert all changes done to this method, this needs to be addressed separately.

File: sdk/src/metrics/data/circular_buffer.cc
Comment (marcalff) [inline sdk/src/metrics/data/circular_buffer.cc:86]: Please explain in more details the change.

Some comments in the code are needed, very hard to understand what it actually does.

Why a while(true) loop, and is there a risk it never ends ?

**Comment by lalitb** [comment]
> Are all the compilers wrong here, is the CI not strict enough, or is clang-tidy wrong ?

I think both are correct, the static analysis just goes a step further:

```bash
/tmp$ echo "#include <stdexcept>

void bar() {
    throw std::runtime_error(\"An error occurred\");
}

void foo() noexcept {
    bar(); 
}

int main() {
    foo();
    return 0;
}" > ./test_exception.cpp
$
$ g++ -Wall -Wextra -Werror -std=c++17 test_exception.cpp -o example
$ clang-tidy test_exception.cpp -checks='-*,bugprone-exception-escape' -- -std=c++17
2 warnings generated.
/tmp/test_exception.cpp:7:6: warning: an exception may be thrown in function 'foo' which should not throw exceptions [bugprone-exception-escape]
void foo() noexcept {
     ^
/tmp/test_exception.cpp:11:5: warning: an exception may be thrown in function 'main' which should not throw exceptions [bugprone-exception-escape]
int main() {
    ^
```

**Review by marcalff — CHANGES_REQUESTED** [review]
Thanks for the cleanup.

This patch contains a lot of very good fixes, and also will improve performances with a lot of object copy that are avoided.

See some changes around CURL, it really needs to use the long type.

About try/catch blocks with exceptions, a bigger discussion is needed with maintainers to see how best to proceed, but given the topic, all the exception related work should probably be done separately.

**Comment by msiddhu** [comment]
> If method foo() is flagged noexcept, and calls method bar() that can raise an exception, I would expect the compiler to raise plenty of errors about that.

`noexcept`  is for compiler optimizations, it's just tell compiler that this method won't throw any exception. But when there is an other function call it may which throws exceptions the function with `noexcept`  will call program to terminate. 

Yes, I feel about exceptions we should discuss more as they would reduce performance (Stack unwinding, Control flow). 

I'll remove the changes related to `try/catch` blocks and create a new PR after the discussion. 

Thanks @marcalff for taking time to review the long PR. Now I understood what I have done wrong. I'll update PR and have comments some comments soon.

File: sdk/src/metrics/data/circular_buffer.cc
Comment (msiddhu) [reply sdk/src/metrics/data/circular_buffer.cc:86]: It's avoid the recursive call. Instead makes an Iterative while loop. 

Anyway reverting this back. Will add in next small PR.

File: examples/etw_threads/main.cc
Comment (msiddhu) [reply examples/etw_threads/main.cc:116]: Noted. Added the **commented parameter names**.

File: exporters/ostream/test/ostream_log_test.cc
Comment (msiddhu) [reply exporters/ostream/test/ostream_log_test.cc:108]: The original code is correct and this aswell. `clang-tidy `warns that there is no comma and the formatting moved them to next line. 

Reverting this.

File: exporters/zipkin/src/zipkin_exporter.cc
Comment (msiddhu) [reply exporters/zipkin/src/zipkin_exporter.cc:78]: Understood. Reverting this.

File: ext/include/opentelemetry/ext/http/client/curl/http_operation_curl.h
Comment (msiddhu) [reply ext/include/opentelemetry/ext/http/client/curl/http_operation_curl.h:271]: Noted and reverted.

File: ext/include/opentelemetry/ext/http/client/curl/http_operation_curl.h
Comment (msiddhu) [reply ext/include/opentelemetry/ext/http/client/curl/http_operation_curl.h:304]: Noted and reverted.

File: ext/src/http/client/curl/http_operation_curl.cc
Comment (msiddhu) [reply ext/src/http/client/curl/http_operation_curl.cc:456]: Noted.

File: ext/src/http/client/curl/http_operation_curl.cc
Comment (msiddhu) [reply ext/src/http/client/curl/http_operation_curl.cc:532]: Done.

File: ext/src/http/client/curl/http_operation_curl.cc
Comment (msiddhu) [reply ext/src/http/client/curl/http_operation_curl.cc:732]: Done

File: ext/src/http/client/curl/http_operation_curl.cc
Comment (msiddhu) [reply ext/src/http/client/curl/http_operation_curl.cc:734]: Done

File: ext/src/http/client/curl/http_operation_curl.cc
Comment (msiddhu) [reply ext/src/http/client/curl/http_operation_curl.cc:776]: Done

File: ext/src/http/client/curl/http_operation_curl.cc
Comment (msiddhu) [reply ext/src/http/client/curl/http_operation_curl.cc:843]: Done

File: sdk/src/logs/logger.cc
Comment (msiddhu) [reply sdk/src/logs/logger.cc:62]: Reverting. We'll take this in next PR

File: exporters/zipkin/src/zipkin_exporter.cc
Comment (marcalff) [inline exporters/zipkin/src/zipkin_exporter.cc:110]: Good find, as both branches in the previous if/else have a return.

Please remove the line then instead of commenting it.

File: sdk/src/metrics/data/circular_buffer.cc
Comment (marcalff) [reply sdk/src/metrics/data/circular_buffer.cc]: Thanks. Fixing all the trivial issues first will reduce the overall noise, so it will be easier to address remaining (and less trivial) fixes later.

**Review by marcalff — APPROVED** [review]
LGTM, thanks for the cleanup.

**Comment by marcalff** [comment]
Waiting for:

* tiny fix in zipskin_exporter.cc
* possible comments from other maintainers

and will merge after that.

File: exporters/zipkin/src/zipkin_exporter.cc
Comment (msiddhu) [reply exporters/zipkin/src/zipkin_exporter.cc]: Removed the comment and return statement.

**Pull request merged**
